### DISCUSSION TITLE
Trimming/AOT - Designs or Constructs that don't work well?

### DISCUSSION THREAD

**Discussion by csturm83** [dotnet/runtime#103064] [Q&A]
Outside of reflection and dynamic code, are there code designs or constructs that make an application more difficult to trim? 

In other words, in the absence of trim warnings, are there general patterns that the trimmer doesn't handle as well that could potentially be avoided (like a general rule of thumb)? 

I understand there are ways to refactor code to help the trimmer out. I'm more wondering if there's a way to generally stay out of its way so it can do its thing reasonably well.

Also, somewhat related, I think I recall in previous .NET releases that some effort went into removing references to LINQ within the BCL to achieve better trimming results. Is such an approach still beneficial to some degree today, or is LINQ more trimmer friendly now?

Thanks in advance for any suggestions. I appreciate all of the hard work that has gone into this tech!

**Comment by huoyaoyuan** [comment]
@dotnet/illink-contrib 

> Also, somewhat related, I think I recall in previous .NET releases that some effort went into removing references to LINQ within the BCL to achieve better trimming results. Is such an approach still beneficial to some degree today, or is LINQ more trimmer friendly now?

LINQ is always trimmer friendly. It's just somehow big.
